src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,124 @@
+        this.poolClosed.set(true);
+        for (PooledWriter<A> writer : this.writers) {
+            writer.close();
+        }
+        this.executor.shutdown();
+    }
+
+    public static class PooledWriter<B> implements WrappedWriter<B> {
+        private final WrappedWriter<B> writer;
+        private final BlockingQueue<B> queue;
+        private final AtomicBoolean writing;
+        private final AtomicBoolean isClosed;
+        public final AsyncWriterPool<B> outer;
+
+        private final int buffsize;
+        private Future<Integer> currentTask;
+
+
+        public PooledWriter(final AsyncWriterPool<B> outer, final WrappedWriter<B> writer, final BlockingQueue<B> queue, final int buffSize) {
+            assert buffSize >= 1; // TODO: Add error message
+            this.writer = writer;
+            this.queue = queue;
+            this.buffsize = buffSize;
+            this.outer = outer;
+            this.writing = new AtomicBoolean(false);
+            this.isClosed = new AtomicBoolean(false);
+            this.currentTask = null;
+
+            // Add self to writer
+            this.outer.writers.add(this);
+        }
+
+        /**
+         * Will throw any error thrown in task
+         */
+        private void checkAndRethrow() {
+            if (!(this.currentTask == null)) {
+                try {
+                    // NB: ignoring output of number of records previously written
+                    this.currentTask.get();
+                } catch (CancellationException | InterruptedException | ExecutionException e) {
+                    this.isClosed.set(true);
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+
+        @Override
+        public void write(final B item) {
+            if (this.isClosed.get()) throw new RuntimeIOException("Attempt to add record to closed writer.");
+
+            // Put new item in queue
+            try {
+                this.queue.put(item);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+
+            // If queue size is large enough, and not currently writing items, send to executor
+            if (queue.size() >= this.buffsize && !this.writing.getAndSet(true)) {
+
+                checkAndRethrow();
+                this.drain(this.buffsize);
+            }
+        }
+
+        private void drain(int toDrain) {
+            // check the result of the previous task
+            this.currentTask = outer.executor.submit(() -> {
+                int counter = 0;
+
+                try {
+                    // BlockingQueue is threadsafe and uses a different lock for both reading and writing
+                    // ends of the queue, so this is safe and avoids an extra coffee.
+                    while (counter < toDrain && !this.queue.isEmpty()) {
+                        this.writer.write(this.queue.poll());
+                        counter++;
```

Comment:
```suggestion
                    B item = null
                    while ((item = this.queue.poll()) != null) {
                        this.writer.write(item);
                        counter++;
```

Reply:
Alternatively I _think_ you could make this more efficient if you did the following (but honestly I'm not 100% sure it would be faster, it all depends on how much the lock acquisition for poll() is):
1. Within each writer define a `private List<B> buffer = new ArrayList<B>(bufferSize)`
2. In the drain method you can then use `this.queue.drainTo(this.buffer, this.bufferSize)` to grab a whole group of items in one operation
3. Then you can write those items to the output and clear the buffer

Reply:
Preliminary attempts with a reused buffer and `drainTo` don't show much speedup.